### DISCUSSION TITLE
add responsive Controls

### DISCUSSION THREAD

**Discussion by lsaavedr** [makeswift/makeswift#779] [Ideas]
Add a parameter to indicate that the control is responsive, for example, in `Marquee` component:

```typescript
      ...
      duration: Number({
        label: 'Animation duration',
        defaultValue: 20,
        suffix: 's',
        responsive: true,
      }),
      ...
```

with this parameter set to `true`, the control can be multi-valued in a responsive way, like the `Style` control. Maybe all controls should always be responsive, just like `Style` control.

**Comment by migueloller** [comment]
@lsaavedr, this is something we've been considering and have been trying to figure out the optimal API. Would love to hear your feedback on this.

Our current plan is to release a new version of `Style` that supports arbitrary values. The API isn't finalized but maybe it will look something like this:

```ts
Style({
  type: Shape({
    color: Color({ label: 'Tint', defaultValue: 'cornflowerblue' })
  }),
  getStyle({ color }) {
    return { backgroundColor: color }
  }
})
```

This would make it so you can do any custom CSS you want and you won't have to worry about responsive values. As for other controls, like `Number`, we can add a `responsive` option but there's a tradeoff here: it requires JavaScript. To be able to provide a `number` to the component, we'll need to use the `matchMedia` API and subscribe to media query changes so we can provide a different value as the viewport resizes. For something like an animation, though, which already requires JS this seems like it would be very helpful. If you ca achieve it with CSS, though, the new version of `Style` would be better to use as you'd get a CSS class.

**Reply by lsaavedr** [reply]
It looks fine, is there a branch where I can checkout this? I'll look...

**Reply by lsaavedr** [reply]
why
```typescipt
import { ResponsiveNumber } from '@makeswift/prop-controllers'
```
is deprecated?

**Reply by migueloller** [reply]
> It looks fine, is there a branch where I can checkout this? I'll look...

As you already found, there's an unstable API that you can try out 😅

This is still WIP, so don't expect it to work nicely out of the box, hence the `unstable_` prefix.

**Reply by migueloller** [reply]
> why
> 
> ```
> import { ResponsiveNumber } from '@makeswift/prop-controllers'
> ```
> 
> is deprecated?

Prop controllers were a legacy method of defining controls internal to Makeswift. They don't compose very well and don't fit in with the new way controls work with List, Shape, etc. That being said, support for responsiveness on our existing controls is something we want to do so that in and of itself _is not_ deprecated—we just haven't built it in to the controls yet.

**Comment by lsaavedr** [comment]
I'm working with unstable_StyleV2, and I have three problems:

1. the parameter in getStyle is undefinied while it's not modified (the default value is expected?)
2. if I build a Shape and change only one parameter of that Shape, all other parameters change to default values ​​in getStyle and not to the expected "tailwind-ordered" value,
3. I still need a "responsive" number, for example to define the size of a nextjs <Image/>, I think that in nextjs <Image/> the size is handled better than in css, or not?

A bit more detail of the second issue is for instance, that for a Shape(width=300(default), height=150(default)), if it's change in desktop with {width:100, height:50} and in laptop it's only change {width:50}, then in laptop the parameter in getStyle is {width:50, height:150(default)} but the parameter expected is {width:50, height:50(from desktop)}.

**Reply by migueloller** [reply]
1. `unstable_StyleV2` doesn't have a default value right now, but since it's a CSS class, `undefined` could be an appropriate value. What were you expecting?
2. I'm not understanding what you mean here. Do you mind providing some more details and maybe some examples?
3. If you need a value like a number, that's not supported by `unstable_StyleV2`, we'll have a separate, `Responsive` control that would allow something like that

**Comment by lsaavedr** [comment]
thanks Miguel,

I have the following code:
```typescript
      iconGapStyle: StyleV2({
        type: Number({
          label: 'Icon gap',
          defaultValue: 2,
          suffix: 'px',
        }),
        getStyle(gap: number) {
          if (gap === undefined) {
            gap = 2
          }

          return { gap: `${gap}px` } as CSSObject
        },
      }),
```
I expected `defaultValue: 2` in `gap` parameter of `getStyle` function, but was `undefined` (with that `if` I can handle that for now).

For the other question, I tried to make a `Shape` like in the stable API:

```typescript
      iconSize: StyleV2({
        type: Shape({
          type: {
            width: Number({
              label: 'Icon width',
              defaultValue: 300,
              suffix: 'px',
            }),
            height: Number({
              label: 'Icon height',
              defaultValue: 150,
              suffix: 'px',
            }),
          }
        }),
        getStyle({ width, height }: { width: number; height: number }) {
          return { width: `${width}px`, height: `${height}px` } as CSSObject
        },
      }),
```
I think that problems are from "undefined default value" also, but in this case it appear in this situation:

1. set width and height in desktop, for example `{with: 100, height: 50}`
2. in external width and height parameters in `getStyle` are as expected `{with: 100, height: 50}`, good!
3. in laptop only change width, for example `{width: 50}`, but now in `getStyle` are `{width: 50, height: 150}` (`150` as `defaultValue`)

regards!

**Reply by migueloller** [reply]
Ah, I see. This would be a bug then as the default value in the Number control should be reflected. We're currently working on some changes that will allow us to consistently apply configurations like default values. @agurtovoy is working on this. Once we've finished that, we will be able to move Style v2 from unstable to stable and properly support default values.

The second problem is also a similar result as what I explain above. In this case, ideally you'd see a width of 50 in laptop and a width of 100 that is inherited from the value set on desktop. Since `getStyle` does always return `height`, though, it should be receiving `defaultValue` of 150 as you're saying. If you did `{ height: height }` without the "px" we could then fall back to desktop for undefined values. It's an interesting phenomenon because it means a default value opts out of any sort of inheritance.
